rs/rust_canisters/random_traffic_test/src/main.rs

```diff
@@ -114,121 +117,184 @@ fn instructions_count_range() -> RangeInclusive<u32> {
+/// still included indirectly.
+fn update_record(response: &api::call::CallResult<Vec<u8>>, index: u32) {
     // Updates the `Reply` at `index` in `RECORDS`.
     let set_reply_in_call_record = move |reply: Reply| {
-        RECORDS.with_borrow_mut(|records| {
-            records[index].reply = Some(reply);
-        });
+        RECORDS.with_borrow_mut(|records| records.get_mut(&index).unwrap().reply = Some(reply));
     };
-
-    let error_code = api::call_with_callbacks(
-        receiver,
-        "handle_call",
-        &vec![0_u8; payload_bytes as usize][..],
-        move || {
-            set_reply_in_call_record(Reply::Bytes(api::arg_data().len() as u32));
-            on_response();
-        },
-        move || {
-            set_reply_in_call_record(Reply::AsynchronousRejection(
-                api::reject_code(),
-                api::reject_message(),
-            ));
-            on_response();
-        },
-    );
-    if error_code == 0 {
-        Ok(())
-    } else {
-        set_reply_in_call_record(Reply::SynchronousRejection(error_code));
-        Err(())
+    match response {
+        Err((_, msg)) if synchronous_rejection(msg) => {
+            // Remove the record for synchronous rejections.
+            SYNCHRONOUS_REJECTIONS_COUNT.set(SYNCHRONOUS_REJECTIONS_COUNT.get() + 1);
+            RECORDS.with_borrow_mut(|records| records.remove(&index).unwrap());
+        }
+        Err((reject_code, msg)) => {
+            set_reply_in_call_record(Reply::Reject(*reject_code as u32, msg.to_string()));
+        }
+        Ok(result) => {
+            set_reply_in_call_record(Reply::Bytes(result.len() as u32));
+        }
     }
 }
 
-/// Samples a weighted binomial distribution to decide whether to make a reply (true) or a
-/// downstream call (false). Defaults to `true` for bad weights (e.g. both 0).
-fn should_reply_now() -> bool {
-    RNG.with_borrow_mut(|rng| {
-        WeightedIndex::new([REPLY_WEIGHT.get(), CALL_WEIGHT.get()])
-            .map_or(true, |dist| dist.sample(rng) == 0)
-    })
+/// Generates `MAX_CALLS_PER_HEARTBEAT` calls as futures. The futures are awaited one by one such
+/// that the records can be updated without awaiting all calls at once.
+#[heartbeat]
+async fn heartbeat() {
+    let (mut futures, mut record_indices) = (Vec::new(), Vec::new());
+    for _ in 0..MAX_CALLS_PER_HEARTBEAT.get() {
+        let Some((future, index)) = setup_call(next_call_tree_id(), 0) else {
+            return;
+        };
+        futures.push(future);
+        record_indices.push(index);
+    }
+
+    while !futures.is_empty() {
+        // Wait for any call to conclude.
+        let (result, index, remaining_futures) = select_all(futures).await;
+
+        // Update records.
+        update_record(&result, record_indices.remove(index));
+
+        // Continue awaiting the remaining futures.
+        futures = remaining_futures;
```

Comment:
```suggestion
        // Wait for any call to conclude.
        let (result, future_index, futures) = select_all(futures).await;

        // Update its record.
        update_record(&result, record_indices.remove(future_index));
```

Reply:
I'm afraid the borrow checker doesn't like this. Something about moving the value in the previous iteration of the loop.